lib/subsequence-provider.js

```diff
@@ -0,0 +1,199 @@
+    if (!prefix) {
+      return
+    }
+
+    if (prefix.trim().length < this.minimumWordLength) {
+      return
+    }
+
+    const buffers = this.includeCompletionsFromAllBuffers
+      ? Array.from(this.watchedBuffers.keys())
+      : [editor.getBuffer()]
+
+    const configSuggestions = this.providerConfig.getSuggestionsForScopeDescriptor(
+      scopeDescriptor
+    )
+
+    const configMatches = this.configSuggestionsToSubsequenceMatches(
+      configSuggestions,
+      prefix
+    )
+
+    const wordsUnderCursors = editor.getCursors().map(cursor =>
+      editor.getBuffer().getTextInRange(cursor.getCurrentWordBufferRange())
+    )
+
+    const subsequenceMatchToType = (match) => {
+      const editor = this.watchedBuffers.get(match.buffer)
+      const scopeDescriptor = editor.scopeDescriptorForBufferPosition(match.positions[0])
+      return this.providerConfig.scopeDescriptorToType(scopeDescriptor)
+    }
+
+    const bufferToSubsequenceMatches = (buffer) => {
+      const assocBuffer = match => { match.buffer = buffer; return match }
+
+      return buffer.findWordsWithSubsequence(
+        prefix,
+        this.additionalWordChars,
+        this.maxResultsPerBuffer,
+      ).then(matches => matches.map(assocBuffer))
+    }
+
+    const sortByScore = matches => matches.sort((a, b) => b.score - a.score)
+
+    const head = matches => _.head(matches, this.maxSuggestions)
+
+    const subsequenceMatchesToSuggestions = matches => matches.map(m => {
+      return m.configSuggestion || {
+        text: m.word,
+        type: subsequenceMatchToType(m)
+      }
+    })
+
+    const uniqueByWord = matches => _.uniq(matches, false, m => m.word)
+
+    const notWordUnderCursor = matches => matches.filter(match =>
+      wordsUnderCursors.indexOf(match.word) === -1 ||
+        match.positions.length > wordsUnderCursors.filter(word => match.word === word).length
+    )
+
+    const applyLocalityBonus = matches => matches.map(match => {
+      if (match.buffer === editor.getBuffer() && match.score > 0) {
+        let lastCursorRow = editor.getLastCursor().getBufferPosition().row
+        let distances = match
+          .positions
+          .map(pos => Math.abs(pos.row - lastCursorRow))
+        let closest = Math.min.apply(Math, distances)
+        match.score += Math.floor(11 / (1 + 0.04 * closest))
+      }
+      return match
+    })
+
+    const removeNonPrefixMatches = matches => matches.filter(match => {
+      const regex = new RegExp('^' + prefix)
+      return regex.exec(match.word)
+    })
+
+    // when thinking about this, go from the bottom (flatten) up
+    const transformToSuggestions = _.compose(
+      subsequenceMatchesToSuggestions,
+      head,
+      notWordUnderCursor,
+      uniqueByWord,
+      sortByScore,
+      (this.useLocalityBonus ? applyLocalityBonus : identity),
+      (this.strictMatching ? removeNonPrefixMatches : identity),
+      _.flatten
+    )
```

Comment:
It's cool to see this logic decomposed in this way, but all of these local closures and `map`, `reduce` and `filter` calls are going to result in a lot of intermediate allocations and garbage collections. I think that for this relatively hot code path, it is worthwhile to avoid that overhead, and you could probably achieve similar readability in a more imperative style.

Some specific thoughts:
* For a given query, a given word should always have the same score. That means that once you sort the matches by score, de-duping the matches can be done cheaply.
* Rather than constructing a complete list of matches and then truncating the list at the end, we can stop our processing once we have found the maximum number of matches.

```js
allMatches.sort((a, b) => b.score - a.score)

const uniqueMatches = []
for (let i = 0, {length} = allMatches; i < length; i++) {
  const match = allMatches[i]

  // Do not include words that are under a cursor
  if (wordsUnderCursor.has(match.word)) continue

  let matchIsDuplicate = false
  for (let j = uniqueMatches.length - 1; j >= 0; j--) {
    const priorMatch = uniqueMatches[j]

    // This condition allows us to de-dup the matches more efficiently than `_.uniq`
    if (priorMatch.score !== match.score) break

    if (priorMatch.word === match.word) {
      matchIsDuplicate = true
      break
    }
  }

  if (!matchIsDuplicate) {
    uniqueMatches.push(match)

    // If there are a huge number of matches, we can avoid some of the work of de-duping them all
    // once we reach our maximum match count
    if (uniqueMatches.length >= this.maxSuggestionCount) break
  }
}
```

Reply:
I couldn't really find evidence that allocations were excessive, but I'm going to change this to a more imperative style anyway.